From 6686296e705c4e8a47f76bc248e97e123930f4db Mon Sep 17 00:00:00 2001
From: Lubomir Rintel <lkundrak@v3.sk>
Date: Tue, 16 Sep 2014 18:22:55 +0200
Subject: [PATCH] ip6: Process RA flags from an unsolicited RA as well

If we see an address we know a RA has arrived, be it a bit early. Set the flags
and proceed to got-ra state.

https://bugzilla.redhat.com/show_bug.cgi?id=1076169
---
 src/ip6-manager/nm-ip6-manager.c |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/src/ip6-manager/nm-ip6-manager.c b/src/ip6-manager/nm-ip6-manager.c
index 5a59578..8e19117 100644
--- a/src/ip6-manager/nm-ip6-manager.c
+++ b/src/ip6-manager/nm-ip6-manager.c
@@ -389,9 +389,10 @@ nm_ip6_device_sync_from_netlink (NMIP6Device *device, gboolean config_changed)
 	nm_log_dbg (LOGD_IP6, "(%s): addresses synced (state %s)",
 	            device->iface, state_to_string (device->state));
 
-	/* We only care about router advertisements if we want a real IPv6 address */
+	/* We got a router advertisement, let's process the flags. If we only see
+	 * an address and not the flag it's been an unsolicited one. */
 	if (   (device->target_state == NM_IP6_DEVICE_GOT_ADDRESS)
-	    && (device->ra_flags & IF_RA_RCVD)) {
+	    && (found_other || device->ra_flags & IF_RA_RCVD)) {
 
 		if (device->state < NM_IP6_DEVICE_GOT_ROUTER_ADVERTISEMENT)
 			device->state = NM_IP6_DEVICE_GOT_ROUTER_ADVERTISEMENT;
-- 
1.7.1

